src/Functions/FunctionMapMapped.h

```diff
@@ -0,0 +1,198 @@
+                            + toString(nested_types.size()) + " arguments. Found "
+                            + arguments[0]->getName() + " instead.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        arguments[0] = std::make_shared<DataTypeFunction>(nested_types);
+    }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        size_t min_args = Impl::needExpression() ? 2 : 1;
+        if (arguments.size() < min_args)
+            throw Exception("Function " + getName() + " needs at least "
+                            + toString(min_args) + " argument; passed "
+                            + toString(arguments.size()) + ".",
+                            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+        if (arguments.size() > 2 && Impl::needOneMap())
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function '{}' needs one map argument", getName());
+
+        const auto * data_type_function = checkAndGetDataType<DataTypeFunction>(arguments[0].type.get());
+
+        if (!data_type_function)
+            throw Exception("First argument for function " + getName() + " must be a function.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        /// The types of the remaining arguments are already checked in getLambdaArgumentTypes.
+
+        DataTypePtr return_type = removeLowCardinality(data_type_function->getReturnType());
+        if (Impl::needBoolean() && !WhichDataType(return_type).isUInt8())
+            throw Exception(
+                "Expression for function " + getName() + " must return UInt8, found " + return_type->getName(),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        const auto * map_type = checkAndGetDataType<DataTypeMap>(arguments[1].type.get());
+        if (!map_type)
+            throw Exception("Second argument for function " + getName() + " must be a map.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        return Impl::getReturnType(return_type, map_type->getKeyValueTypes());
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t /*input_rows_count*/) const override
+    {
+        const auto & column_with_type_and_name = arguments[0];
+
+        if (!column_with_type_and_name.column)
+            throw Exception("First argument for function " + getName() + " must be a function.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        const auto * column_function = typeid_cast<const ColumnFunction *>(column_with_type_and_name.column.get());
+
+        if (!column_function)
+            throw Exception("First argument for function " + getName() + " must be a function.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        ColumnPtr offsets_column;
+
+        ColumnPtr column_first_map_ptr;
+        const ColumnMap * column_first_map = nullptr;
+
+        ColumnsWithTypeAndName maps;
```

Comment:
What's `maps` for? It and code below looks like logical unit, we can move it into separate function and it would be clear by the name. Or add a comment why do we perform such transformation